File: py/src/aoc24/day9.py
import enum
from typing import Self
from dataclasses import dataclass
from aoc24 import utils


SimpleMemory = list[int | None]

FREE_ID = -1


@dataclass
class Block:
    length: int
    id_: int

    @property
    def is_free(self) -> bool:
        return self.id_ == FREE_ID

    def checksum(self, start: int) -> int:
        if self.is_free:
            return 0

        return sum(i * self.id_ for i in range(start, start + self.length))

    def copy(self):
        return Block(self.length, self.id_)

    @classmethod
    def free(cls, length: int):
        return Block(length, FREE_ID)


@dataclass
class AdvancedMemory:
    blocks: list[Block]
    highest_id: int

    @classmethod
    def from_txt(cls, inp: str) -> Self:
        inp = inp.rstrip()
        curr_id = 0
        blocks: list[Block] = []
        is_file = True

        for x in inp:
            times = int(x)

            if is_file:
                block = Block(times, curr_id)
                curr_id += 1
            else:
                block = Block(times, -1)

            blocks.append(block)
            is_file = not is_file

        return cls(blocks, curr_id - 1)

    def copy(self):
        blocks = [x.copy() for x in self.blocks]

        return AdvancedMemory(blocks, self.highest_id)

    def __iter__(self):
        return iter(self.blocks)

    def __len__(self):
        return len(self.blocks)

    def __getitem__(self, idx: int):
        return self.blocks[idx]

    def __setitem__(self, idx: int, val: Block) -> None:
        self.blocks[idx] = val


def to_simple_memory(inp: str) -> SimpleMemory:
    inp = inp.rstrip()
    curr_id = 0
    res: SimpleMemory = []
    is_file = True

    for x in inp:
        times = int(x)

        if is_file:
            res.extend([curr_id] * times)
            curr_id += 1
        else:
            res.extend([None] * times)
        is_file = not is_file

    return res


def simple_compact(memory: SimpleMemory) -> SimpleMemory:
    mem: SimpleMemory = memory[:]

    free_slot = mem.index(None)
    to_move_idx = len(mem) - 1

    while to_move_idx > free_slot and mem[to_move_idx] is None:
        to_move_idx -= 1

    while free_slot < to_move_idx:
        mem[free_slot] = mem[to_move_idx]
        mem[to_move_idx] = None

        while free_slot < to_move_idx and mem[free_slot] is not None:
            free_slot += 1

        while to_move_idx > free_slot and mem[to_move_idx] is None:
            to_move_idx -= 1

    return mem


def simple_checksum(mem: SimpleMemory) -> int:
    return sum(i * x for (i, x) in enumerate(mem) if x is not None)


def adv_checksum(mem: AdvancedMemory) -> int:
    checksum = 0

    offset = 0
    for block in mem:
        checksum += block.checksum(offset)
        offset += block.length

    return checksum


def part1():
    txt = utils.read_contents(9)
    memory = to_simple_memory(txt)
    compact = simple_compact(memory)
    res = simple_checksum(compact)

    print(f"Day 9 part 1: {res}")


def advanced_compact(memory: AdvancedMemory) -> AdvancedMemory:
    mem = memory.copy()

    to_move_id = memory.highest_id
    prev_to_move_idx = len(mem)
    while to_move_id >= 0:
        to_move_idx = -1
        for i in range(prev_to_move_idx):
            if mem[i].id_ == to_move_id:
                to_move_idx = i
                break
        if to_move_idx == -1:
            to_move_id -= 1
            continue

        prev_to_move_idx = to_move_idx
        block_to_move = mem[to_move_idx]
        required_space = block_to_move.length
        free_slot = -1
        for slot, block in enumerate(mem.blocks[:to_move_idx]):
            if block.is_free and block.length >= required_space:
                free_slot = slot
                break

        if free_slot != -1:
            # eq space or less space
            free_block = mem[free_slot]
            if free_block.length == required_space:
                # simple swap is okay
                mem[free_slot] = block_to_move
                mem[to_move_idx] = free_block
            else:
                # split current free into into two parts
                mem[to_move_idx] = Block.free(required_space)
                mem[free_slot] = Block(required_space, block_to_move.id_)
                mem.blocks.insert(free_slot + 1, Block.free(free_block.length - required_space))

        to_move_id -= 1

    return mem


def part2():
    txt = utils.read_contents(9)
    memory = AdvancedMemory.from_txt(txt)
    compact = advanced_compact(memory)
    res = adv_checksum(compact)

    print(f"Day 9 part 2: {res}")


if __name__ == "__main__":
    test1 = "12345"
    test2 = "2333133121414131402"

    test2_simple_mem = to_simple_memory(test2)
    test2_adv_mem = AdvancedMemory.from_txt(test2)

    test2_simple_comp = simple_compact(test2_simple_mem)
    test2_adv_comp = advanced_compact(test2_adv_mem)

    print(simple_checksum(test2_simple_comp))
    print(adv_checksum(test2_adv_comp))

File: oxide/src/day9.rs
use crate::utils;

type Num = u64;
type SimpleBlock = Option<Num>;
type SimpleMemory = Vec<SimpleBlock>;

#[derive(Clone, Copy)]
struct AdvBlock {
    length: u32,
    val: SimpleBlock,
}

impl AdvBlock {
    fn new(length: u32, val: SimpleBlock) -> Self {
        Self { length, val }
    }

    fn is_free(&self) -> bool {
        self.val.is_none()
    }

    fn free_with_len(length: u32) -> Self {
        Self::new(length, Option::None)
    }

    fn checksum(&self, start: Num) -> Num {
        match self.val {
            None => 0,
            Some(v) => (start..start + self.length as Num).map(|s| s * v).sum(),
        }
    }

    fn split(&self, new_size: u32, block: SimpleBlock) -> (Self, Option<Self>) {
        let a = AdvBlock::new(new_size, block);
        let b = if new_size == self.length {
            Option::None
        } else {
            Option::Some(Self::free_with_len(self.length - new_size))
        };

        (a, b)
    }
}

fn load_simple_data() -> SimpleMemory {
    let txt = utils::read_day_file(9);

    let mut curr_id: Num = 0;
    let mut is_file = true;

    let mut blocks: SimpleMemory = Vec::new();

    for ch in txt.trim_end().chars() {
        let times = ch.to_digit(10).unwrap() as usize;

        let block: SimpleBlock = if is_file {
            curr_id += 1;
            Option::Some(curr_id - 1)
        } else {
            Option::None
        };

        let slice = std::iter::repeat_n(block, times).collect::<Vec<_>>();
        blocks.extend_from_slice(&slice);
        is_file = !is_file;
    }

    blocks
}

fn simple_compress(memory: &SimpleMemory) -> SimpleMemory {
    let mut mem = memory.clone();

    let mut free_slot = 0;
    let mut to_move_idx = mem.len() - 1;

    while free_slot < to_move_idx && mem[free_slot].is_some() {
        free_slot += 1;
    }

    while to_move_idx > free_slot && mem[to_move_idx].is_none() {
        to_move_idx -= 1;
    }

    while free_slot < to_move_idx {
        mem[free_slot] = mem[to_move_idx];
        mem[to_move_idx] = Option::None;

        while free_slot < to_move_idx && mem[free_slot].is_some() {
            free_slot += 1;
        }

        while to_move_idx > free_slot && mem[to_move_idx].is_none() {
            to_move_idx -= 1;
        }
    }

    mem
}

fn simple_checksum(memory: &SimpleMemory) -> usize {
    memory
        .iter()
        .enumerate()
        .map(|(pos, x)| pos * x.unwrap_or(0) as usize)
        .sum::<usize>()
}

pub fn part1() {
    let mem: SimpleMemory = load_simple_data();
    let compressed = simple_compress(&mem);
    let checksum = simple_checksum(&compressed);

    println!("Day 9 part 1: {checksum}")
}

type AdvMemory = Vec<AdvBlock>;

fn load_adv_data() -> (AdvMemory, Num) {
    let txt = utils::read_day_file(9);

    let mut curr_id: Num = 0;
    let mut is_file = true;

    let mut blocks: AdvMemory = Vec::with_capacity(txt.len());

    for ch in txt.trim_end().chars() {
        let times = ch.to_digit(10).unwrap();

        let block: AdvBlock = if is_file {
            curr_id += 1;

            AdvBlock::new(times, Option::Some(curr_id - 1))
        } else {
            AdvBlock::free_with_len(times)
        };

        blocks.push(block);
        is_file = !is_file;
    }

    (blocks, curr_id - 1)
}

fn adv_compress(memory: &AdvMemory, highest_id: Num) -> AdvMemory {
    let mut mem = memory.clone();

    let mut to_move_id = highest_id;
    let total_blocks = mem.len() - 1;
    let mut prev_to_move_idx = total_blocks + 1;

    while to_move_id > 0 {
        let mut to_move_idx = Option::None;

        for (i, block) in mem[..prev_to_move_idx].iter().rev().enumerate() {
            if let Some(id) = block.val {
                if id == to_move_id {
                    to_move_idx = Option::Some(total_blocks - i);
                    break;
                }
            }
        }
        if to_move_idx.is_none() {
            to_move_id -= 1;
            continue;
        }

        // Found a block to move, now find a block with enough memory
        prev_to_move_idx = to_move_idx.unwrap();
        let block_to_move = mem[prev_to_move_idx];
        let required_size = block_to_move.length;

        let mut free_slot = Option::None;
        for (i, block) in mem[..prev_to_move_idx].iter().enumerate() {
            if block.is_free() && block.length >= required_size {
                free_slot = Option::Some(i);
                break;
            }
        }

        if let Some(free_slot_idx) = free_slot {
            // Two cases, equal size or more
            let (a, b) = mem[free_slot_idx].split(required_size, block_to_move.val);
            mem[prev_to_move_idx] = AdvBlock::free_with_len(required_size);
            mem[free_slot_idx] = a;

            if let Some(another_free) = b {
                mem.insert(free_slot_idx + 1, another_free);
            }
        }

        to_move_id -= 1;
    }

    mem.to_vec()
}

fn adv_checksum(memory: &AdvMemory) -> Num {
    let mut checksum: Num = 0;
    let mut offset: Num = 0;

    for block in memory {
        checksum += block.checksum(offset);
        offset += block.length as Num;
    }

    checksum
}

pub fn part2() {
    let (mem, highest_id) = load_adv_data();
    let compressed = adv_compress(&mem, highest_id);
    let checksum = adv_checksum(&compressed);

    println!("Day 9 part 2: {checksum}")
}
